### ISSUE TITLE
Types are broken with `compilerOptions.module` set to `nodenext` due to imports missing file extensions

### ISSUE THREAD

**Issue by jaydenseric** [apollographql/apollo-client#9890]
**Intended outcome:**

Projects with TypeScript config `compilerOptions.module` set to `nodenext` or `node16` should be able to import `@apollo/client` modules with working types. This is critical for using TypeScript correctly with modern Node.js packages or projects that use the package `exports` field, ESM in `.mjs` files, etc.

**Actual outcome:**

Types are broken for `@apollo/client` modules that have types that import other modules/types without correctly specifying the file extensions in the imports. 

**How to reproduce the issue:**

In `package.json`:

```json
{
  "name": "demo",
  "private": true,
  "dependencies": {
    "@apollo/client": "^3.6.9"
  }
}
```

In `jsconfig.json`:

```json
{
  "compilerOptions": {
    "maxNodeModuleJsDepth": 10,
    "module": "nodenext",
    "noEmit": true,
    "strict": true
  }
}
```

In `demo.mjs`:

```js
// @ts-check

import { ApolloLink } from "@apollo/client/link/core/ApolloLink.js";

new ApolloLink((request) => {
  console.log(request);
})
```

Notice how the types are incomplete for the `ApolloLink` constructor arguments and fall back to `any`:

<img width="1082" alt="Screen Shot 2022-07-07 at 6 29 01 pm" src="https://user-images.githubusercontent.com/1754873/177731137-2c3f90f6-7787-4ca9-9b6d-73ae6a43bd32.png">

If you dig into the types for it, you can see that the `RequestHandler` type is being imported incorrectly (missing the file extension) from `./types`, and TypeScript is unable to tell what its type is:

<img width="1389" alt="Screen Shot 2022-07-07 at 6 30 23 pm" src="https://user-images.githubusercontent.com/1754873/177731185-7046aae5-ef56-430e-bbbd-44f686197806.png">

If you change the import path to `./types.js` (add the missing file extension), the types start to work:

<img width="1389" alt="Screen Shot 2022-07-07 at 6 30 34 pm" src="https://user-images.githubusercontent.com/1754873/177731479-aa7639d6-fefd-4790-8c48-7540ba2f975d.png">

Ultimately fixing the types where `ApolloLink` is being constructed in the project:

<img width="1389" alt="Screen Shot 2022-07-07 at 6 31 09 pm" src="https://user-images.githubusercontent.com/1754873/177731674-3d3a86b0-6341-49f3-b5ef-f9e86b5be206.png">

All imports, including for type only imports, should specify the full path including the file name for the module being imported.

As a side note, type only imports should use the `type` keyword but currently they don't:

https://github.com/apollographql/apollo-client/blob/289336a5e126cb4b4ca9ec08badc68c9cd3e50fd/src/link/core/ApolloLink.ts#L4-L10

E.g:

```ts
import type {
  NextLink,
  Operation,
  RequestHandler,
  FetchResult,
  GraphQLRequest
} from './types.js';
```

**Versions**

- Node.js: v18.4.0
- `typescript`: v4.7.4
- `@apollo/client`: v3.6.9

**Comment by jaydenseric** [comment]
This issue is blocking a big update I'm working on for [`apollo-upload-client`](https://github.com/jaydenseric/apollo-upload-client), which (among other things) adds TypeScript types for the package.

**Comment by trevor-scheer** [comment]
We were able to resolve this same issue for AS4 via https://github.com/apollographql/apollo-server/pull/6731

**Comment by GeenCo86** [comment]
> We were able to resolve this same issue for AS4 via [apollographql/apollo-server#6731](https://github.com/apollographql/apollo-server/pull/6731)

What is the result of this. Is there a fix in the client library or will it ?

**Comment by trevor-scheer** [comment]
The outcome of that PR is that this same issue was resolved for Apollo _Server_ in our recent v4 release. Prioritizing and implementing a fix for this in Apollo Client is up to the @apollographql/client-typescript team. I've just provided the PR as additional context.

It's likely they'd be interested in reviewing a PR which provably (i.e. with tests) resolves the issue without breaking existing use cases!

**Comment by shaunco** [comment]
Any updates on this? This is causing us quite a bit of pain as we try to move from CJS to ESM, and it seems like a pretty simple fix for the Apollo client...

**Comment by vtereshyn** [comment]
would be great if you guys provide an update here. thank you in advance!

**Comment by jerelmiller** [comment]
@phryneas would you mind weighing in here? I'm not sure if this type of change would be considered breaking or whether we can do a minor here.

**Comment by phryneas** [comment]
I *think* it might have broader implications, like us also needing to ship `.d.cts` files etc., but I'm not exactly sure.

@markerikson, I would be very thankful if you could chime in here.

**Comment by phryneas** [comment]
@markerikson  Unfortunately, exactly that one - concatenating all imports with `.js` and `/index.js`. We have non-rolled up artifacts, both in types and runtime code. But with my luck, of course that's the one thing you didn't have to fight yet ^^ Thanks for taking a look :)

**Comment by jaydenseric** [comment]
@markerikson

> FWIW, I think this is why I've opted to bundle all of our artifacts for the Redux libs

Never bundle modules that are published, that's a big anti-pattern. It forces users to import and load everything in your library regardless if they are using it or not. Packages with multiple public exports should never have an index module; users should deep import only the export they need from a specific module that has a single export. You can learn more about optimal module design here:

https://jaydenseric.com/blog/optimal-javascript-module-design

**Comment by markerikson** [comment]
@jaydenseric we've been shipping bundles for most of our libraries for years, and I'm _especially_ doing that now that we're trying to ship full ESM/CJS compat to avoid other pain points.

Tree shaking seems to work fine with Vite/ESBuild, Next/Webpack, etc.  Just last night I was converting React-Redux to have bundled artifacts instead of separate transpiled JS files on disk, and confirmed that a Vite app correctly tree-shakes `connect` from an app bundle if it's not being used.

Note that we assume that almost all of our users are building apps using a bundler, not importing raw ESM into a browser.  Along with that, we have embedded `process.env.NODE_ENV` checks for dev vs prod behavior in our source code.  (We _are_ shipping a browser-friendly ESM artifact that has those `NODE_ENV` checks compiled away in the next majors, specifically to support the "load ESM into a browser" use case.)

(FWIW, I just finished reading your article, and while I appreciate the amount of time you put into writing it, I generally disagree with most of the points being made and don't intend to re-change the packaging setups I've just put months into updating.)

**Comment by phryneas** [comment]
@jaydenseric could you give `@apollo/client@0.0.0-pr-10994-20230623115940` a spin? (see #10994)

**Comment by wootencl** [comment]
@phryneas A little late to the party here but super appreciative for tackling this issue 🙏 . In the midst of converting my org's libraries to ESM and this was definitely an unexpected gotcha. Can confirm `~3.8.0` fixes the TS issues on my end and believe this issue can be closed. 

/cc @jaydenseric

**Comment by phryneas** [comment]
Ah right, I forgot to close this after we shipped 3.8 - good call, closing :)
